accepted/2021/preview-features/preview-features.md

```diff
@@ -0,0 +1,418 @@
+behavioral changes and also allows certain parts of .NET to make deliberate
+breaking changes if necessary.
+
+As a result, this makes it viable to innovate across runtime, libraries, and
+languages, allowing us to build more compelling features that would have been
+almost impossible to do on .NET Framework. Examples include static linking,
+default interface members, and ref structs (such as `Span<T>`).
+
+However, it's sometimes challenging to deliver cross-cutting feature work in a
+single release, especially when they are breaking new ground in terms of
+expressiveness. A good example of such a feature is the upcoming improved
+numeric support in generic code, with the help of static interface members. Such
+a feature requires work in the runtimes, the languages, and the libraries. The
+work at the library level isn't very high, but it's all about finding the right
+set of support infrastructure so that developers are able to be productive at
+writing generic code that can rely on numerical expressions and operators. In a
+sense, the library work can't really fully start until the runtime and compiler
+support is sufficiently bootstrapped, but it's the most important part in the
+validation of the feature set. Practically, this means a good chunk of the
+feature will only be available halfway through a release cycle, which greatly
+reduces the time window for gathering feedback from other parts of the product
+as well as from customers using preview builds of the upcoming .NET version.
+
+Sadly, preview feedback isn't always sufficient to judge how well a feature
+works because
+
+1. Many fewer people use preview builds of .NET and
+2. Many people don't use .NET previews for their production projects, thus
+   greatly reducing the breadth & depth at which previews are being exercised.
+
+This document is exploring a model where GA quality builds of .NET can contain
+features that are considered still in preview.
+
+The goal is to ensure that customers can discover those features & have an easy
+way to play with them while simultaneously being fully aware that these features
+aren't supported yet and thus might change (or be removed entirely) between
+versions of .NET, just like they could in preview builds.
+
+## Scenarios and User Experience
+
+### Enabling a preview runtime feature
+
+Maddy is building a Tye application. She reads in a blog post that the .NET
+runtime team is experimenting with a new GC mode, code named FancyGC.
+
+She wants to see how this improves the performance of her backend. She's glad to
+see that the new GC mode was shipped as part of .NET 6, so she doesn't need to
+figure out how to run a new version of .NET in her CI lab. So she creates a
+feature branch and enables the new GC mode simply by changing the service's
+project file as follows:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <EnablePreviewFeatures>True</EnablePreviewFeatures>
```

Comment:
Where does this example says that it enables `FancyGC`?

The preview runtime features that are just performance optimizations should be on a different plan, and it should be possible to enable them indendently.

Reply:
I'm not opposed to that, but my understanding from the previous meeting was that we didn't want a model where individual features can be turned on/off but instead we have a single preview mode that turns on all preview features together.

Reply:
For runtime experiments that change behavior without requiring or in any way relying on additional APIs or language features, we're fine being on the same plan we've been on... a bazillion environment variables for configuring lots of things, some of which are experimental.  I would prefer if we started naming them DOTNET_ instead of COMPlus_ (https://github.com/dotnet/runtime/issues/47283), but that's orthogonal :smile:

Reply:
> we didn't want a model where individual features can be turned on/off 

That is correct for functionality features: New APIs, new language features, etc. 

I expect that performance oriented features with no public surface are typically going to be different. Notice the shipping version of the performance oriented features are often opt-in as well. In your Fancy GC example, the shipping version of it would be most likely opt-in as well, e.g. same as ServerGC is opt-in today.

Some of the performance oriented features can be an exception from this rule and it may make sense to activate them automatically in preview mode - if we expect that the eventual shipping version of the performance oriented feature is going to be on by default as well.

Some of the performance oriented features affect functionality. A good example is IL linking. We should never auto activate features like that in preview mode.

Reply:
We addressed this by scoping out pure runtime features from this.